src/main/java/Dude/Dude.java

```diff
@@ -0,0 +1,601 @@
+            String endDate = event.end.format(formatter);
+            sb.append("E|").append(task.isDone ? "1" : "0").append("|").append(task.description).append("|").append(startDate).append("|").append(endDate);
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Converts a string representation of a task back into a Task object.
+     * The string format should include task type, completion status, description, and dates if applicable.
+     *
+     * @param line The string representation of the task.
+     * @return A Task object corresponding to the string, or null if the string format is invalid.
+     */
+    private static Task taskFromString(String line) {
+        String[] parts = line.split("\\|");
+        boolean isDone = "1".equals(parts[1]);
+        String description = parts[2];
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+        try {
+            switch (parts[0]) {
+            case "T":
+                ToDo todo = new ToDo(description);
+                if (isDone) todo.markAsDone();
+                return todo;
+            case "D":
+                Deadline deadline = new Deadline(description, parts[3]);
+                if (isDone) deadline.markAsDone();
+                return deadline;
+            case "E":
+                Event event = new Event(description, parts[3], parts[4]);
+                if (isDone) event.markAsDone();
+                return event;
+            default:
+                return null;
+            }
+        } catch (DateTimeParseException e) {
+            System.out.println("Error parsing date: " + e.getMessage());
+            return null;
+        }
+    }
+}
+
+class Ui {
+    private Scanner scanner;
+
+    public Ui() {
+        this.scanner = new Scanner(System.in);
+    }
+
+    public void showWelcome() {
+        System.out.println("Hello! I'm Dude\nWhat can I do for you?");
+    }
+
+    public void showAddTask(Event event){
+        System.out.println("Added Event: " + event);
+    }
+
+    public void showAddTask(Deadline deadline) {
+        System.out.println("Added Deadline: " + deadline);
+    }
+
+    public void showAddTask(ToDo todo) {
+        System.out.println("Added Todo: " + todo);
+    }
+
+    public void showDelete(Task task) {
+        System.out.println("Deleted: "  + task);
+    }
+
+    public void showDone(Task task) {
+        System.out.println("Marked as Done: " + task);
+    }
+
+    public void showUndone(Task task) {
+        System.out.println("Marked as Not Done: " + task);
+    }
+
+
+    public void showGoodbye() {
+        System.out.println("Bye. Hope to see you again soon!");
+    }
+
+    public void showError(String message) {
+        System.out.println("Error: " + message);
+    }
+
+    public void showMessage(String message) {
+        System.out.println(message);
+    }
+
+    public void showTaskList(ArrayList<Task> tasks) {
```

Comment:
I like that other than following naming conventions, you use the keyword "show" for all Ui messages, this makes following using the methods easy with minimal information overloading.